bpf/cpu/cpu.bpf.c

```diff
@@ -252,117 +268,102 @@ bpf_map_lookup_or_try_init(void *map, const void *key, const void *init) {
-  // left:%d, right:%d)", ctx->table->rows[mid].pc, ctx->pc, index,
-  // MAX_BINARY_SEARCH_DEPTH, mid, ctx->left, ctx->right);
-  if (ctx->table->rows[mid].pc <= ctx->pc) {
-    ctx->found = mid;
-    ctx->left = mid + 1;
-  } else {
-    ctx->right = mid;
-  }
+    // Appease the verifier.
+    if (mid < 0 || mid >= MAX_UNWIND_TABLE_SIZE) {
+      bpf_printk("\t.should never happen");
+      BUMP_UNWIND_SHOULD_NEVER_HAPPEN_ERROR();
+      return BINARY_SEARCH_SHOULD_NEVER_HAPPEN;
+    }
 
-  // Debug logs.
-  // bpf_printk("\t<- fetched PC %llx, target PC %llx (iteration %d/%d, mid: --,
-  // left:%d, right:%d)", ctx->table->rows[mid].pc, ctx->pc, index,
-  // MAX_BINARY_SEARCH_DEPTH, ctx->left, ctx->right);
+    // Debug logs.
+    // bpf_printk("\t-> fetched PC %llx, target PC %llx (iteration %d/%d, mid:
+    // %d, left:%d, right:%d)", ctx->table->rows[mid].pc, ctx->pc, index,
+    // MAX_BINARY_SEARCH_DEPTH, mid, ctx->left, ctx->right);
+    if (table->rows[mid].pc <= pc) {
+      found = mid;
+      left = mid + 1;
+    } else {
+      right = mid;
+    }
 
-  return 0;
+    // Debug logs.
+    // bpf_printk("\t<- fetched PC %llx, target PC %llx (iteration %d/%d, mid:
+    // --, left:%d, right:%d)", ctx->table->rows[mid].pc, ctx->pc, index,
+    // MAX_BINARY_SEARCH_DEPTH, ctx->left, ctx->right);
+  }
+  return BINARY_SEARCH_EXHAUSTED_ITERATIONS;
 }
 
 // Print an unwinding table row for debugging.
 static __always_inline void show_row(stack_unwind_table_t *unwind_table,
                                      int index) {
   u64 pc = unwind_table->rows[index].pc;
-  int cfa_reg = unwind_table->rows[index].cfa_reg;
-  int cfa_offset = unwind_table->rows[index].cfa_offset;
-  int rbp_offset = unwind_table->rows[index].rbp_offset;
+  u16 cfa_type = unwind_table->rows[index].cfa_type;
+  s16 cfa_offset = unwind_table->rows[index].cfa_offset;
+  s16 rbp_offset = unwind_table->rows[index].rbp_offset;
 
   bpf_printk("~ %d entry. Loc: %llx, CFA reg: %d Offset: %d, $rbp %d", index,
-             pc, cfa_reg, cfa_offset, rbp_offset);
+             pc, cfa_type, cfa_offset, rbp_offset);
 }
-static __always_inline int
-walk_user_stacktrace(bpf_user_pt_regs_t *regs,
-                     stack_unwind_table_t *unwind_table, stack_trace_t *stack) {
-  u64 current_rip = regs->ip;
-  u64 current_rsp = regs->sp;
-  u64 current_rbp = regs->bp;
-
-  bpf_printk("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
-  bpf_printk("traversing stack using .eh_frame information!!");
-  bpf_printk("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
-
-  u64 table_len = unwind_table->table_len;
-  // Just for debugging to ensure that the data we are reading
-  // matches what we wrote.
-  bpf_printk("- unwind table has %d items", table_len);
 
-  // Invariant check.
-  if (table_len >= MAX_UNWIND_TABLE_SIZE) {
-    bpf_printk("should never happen");
-    BUMP_UNWIND_SHOULD_NEVER_HAPPEN_ERROR();
+SEC("perf_event")
```

Comment:
Why is this needed? I always thought it was for hooking to the event. Does it also use for exporting functions?

Reply:
Yes, all the other functions are inlined and one single program is executed. The compiler has replaced every call site with its body and they don't really exist as functions or individual programs anymore. 

As we are doing a tail call to a function, this has to be present and has to be the same BPF type as the BPF program that invokes it